mantle/kola/harness.go

```diff
@@ -436,6 +448,153 @@ func RunUpgradeTests(patterns []string, pltfrm, outputDir string, propagateTestE
+	remotepath := fmt.Sprintf("/usr/local/bin/kola-runext-%s", filepath.Base(executable))
+	unit := fmt.Sprintf(`[Unit]
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+Environment=%s=%s
+ExecStart=%s
+[Install]
+RequiredBy=multi-user.target
+`, kolaExtBinDataEnv, kolaExtBinDataDir, remotepath)
+	runextconfig := ignv3types.Config{
+		Ignition: ignv3types.Ignition{
+			Version: "3.0.0",
+		},
+		Systemd: ignv3types.Systemd{
+			Units: []ignv3types.Unit{
+				{
+					Name:     unitname,
+					Contents: &unit,
+					Enabled:  util.BoolToPtr(false),
+				},
+			},
+		},
+	}
+
+	finalIgn := ignv3.Merge(ignc3, runextconfig)
+	serializedIgn, err := json.Marshal(finalIgn)
+	if err != nil {
+		return errors.Wrapf(err, "serializing ignition")
+	}
+
+	register.RegisterTest(&register.Test{
+		Name:          testname,
+		ClusterSize:   1, // Hardcoded for now
+		ExternalTest:  executable,
+		DependencyDir: dependencydir,
+
+		Run: func(c cluster.TestCluster) {
+			mach := c.Machines()[0]
+			plog.Debugf("Running kolet")
+			_, stderr, err := mach.SSH(fmt.Sprintf("sudo ./kolet run-test-unit %s", shellquote.Join(unitname)))
+			out, _, suberr := mach.SSH(fmt.Sprintf("sudo systemctl status %s", shellquote.Join(unitname)))
+			if suberr != nil {
+				fmt.Printf("systemctl status %s:\n%s\n", unitname, string(out))
+			}
+			if err != nil {
+				if Options.SSHOnTestFailure {
+					plog.Errorf("dropping to shell: kolet failed: %v: %s", err, stderr)
+					platform.Manhole(mach)
+				} else {
+					c.Fatalf(errors.Wrapf(err, "kolet failed: %s", stderr).Error())
+				}
+			}
+		},
+
+		UserDataV3: conf.Ignition(string(serializedIgn)),
+	})
+
+	return nil
+}
+
+// registerTestDir parses one test directory and registers it as a test
+func registerTestDir(dir, testprefix string, children []os.FileInfo) error {
+	var dependencydir string
+	ignition := `{ "ignition": { "version": "3.0.0" } }`
+	executables := []string{}
+
+	for _, c := range children {
+		isreg := c.Mode().IsRegular()
+		if isreg && (c.Mode().Perm()&0001) > 0 {
+			executables = append(executables, filepath.Join(dir, c.Name()))
+		} else if isreg && c.Name() == "config.ign" {
+			v, err := ioutil.ReadFile(filepath.Join(dir, c.Name()))
+			if err != nil {
+				return errors.Wrapf(err, "reading %s", c.Name())
+			}
+			ignition = string(v)
+		} else if c.IsDir() && c.Name() == "deps" {
+			dependencydir = filepath.Join(dir, c.Name())
+		} else if c.Mode()&os.ModeSymlink != 0 && c.Name() == "deps" {
+			target, err := filepath.EvalSymlinks(filepath.Join(dir, c.Name()))
+			if err != nil {
+				return err
+			}
+			dependencydir = target
+		} else if c.IsDir() {
+			subdir := filepath.Join(dir, c.Name())
+			subchildren, err := ioutil.ReadDir(subdir)
+			if err != nil {
+				return err
+			}
+			subprefix := fmt.Sprintf("%s.%s", testprefix, c.Name())
+			if err := registerTestDir(subdir, subprefix, subchildren); err != nil {
+				return err
+			}
+		}
+	}
+
+	if len(executables) == 1 {
```

Comment:
Do you think that putting everything in a loop and just special casing the name is more readable, i.e.:

```golang
for _, executable := range executables {
    testname := testprefix
    if len(executables) > 1 {
        testname = fmt.Sprintf("%s.%s", testname, filepath.Base(executable))
    }
    err := registerExternalTest(testname, executable, dependencydir, ignition)
    if err != nil {
        return err
    }
}
```